src/sentry/tasks/assemble.py

```diff
@@ -251,53 +251,68 @@ def _normalize_debug_id(debug_id: Optional[str]) -> Optional[str]:
         return None
 
 
-def _extract_debug_ids_from_manifest(manifest: dict) -> List[Tuple[SourceFileType, str]]:
+def _extract_debug_ids_from_manifest(
+    manifest: dict,
+) -> Tuple[Optional[str], List[Tuple[SourceFileType, str]]]:
     debug_ids_with_types = []
 
+    # We also want to extract the bundle_id which is also known as the bundle debug_id. This id is used to uniquely
+    # identify a specific ArtifactBundle in case for example of future deletion.
+    #
+    # If no id is found, it means that we must have an associated release to this ArtifactBundle, through the
+    # ReleaseArtifactBundle table.
+    bundle_id = manifest.get("debug_id")
+    if bundle_id is not None:
+        bundle_id = _normalize_debug_id(bundle_id)
+
     files = manifest.get("files", {})
     for file_path, info in files.items():
         headers = _normalize_headers(info.get("headers", {}))
-        if (debug_id := headers.get("debug-id", None)) is not None:
+        if (debug_id := headers.get("debug-id")) is not None:
             debug_id = _normalize_debug_id(debug_id)
-            file_type = info.get("type", None)
+            file_type = info.get("type")
             if (
                 debug_id is not None
                 and file_type is not None
                 and (source_file_type := SourceFileType.from_lowercase_key(file_type)) is not None
             ):
                 debug_ids_with_types.append((source_file_type, debug_id))
 
-    return debug_ids_with_types
+    return bundle_id, debug_ids_with_types
 
 
 def _create_artifact_bundle(
-    release: Optional[Release],
-    dist: Optional[Distribution],
+    version: Optional[str],
+    dist: Optional[str],
     org_id: int,
-    project_ids: List[int],
+    project_ids: Optional[List[int]],
     archive_file: File,
     artifact_count: int,
-) -> bool:
+):
     with ReleaseArchive(archive_file.getfile()) as archive:
-        debug_ids_with_types = _extract_debug_ids_from_manifest(archive.manifest)
+        bundle_id, debug_ids_with_types = _extract_debug_ids_from_manifest(archive.manifest)
 
-        if len(debug_ids_with_types) > 0:
+        # We want to save an artifact bundle only if we have found debug ids in the manifest or if the user specified
+        # a release for the upload.
+        if len(debug_ids_with_types) > 0 or version is not None:
             artifact_bundle = ArtifactBundle.objects.create(
                 organization_id=org_id,
-                bundle_id=uuid.uuid4().hex,
+                bundle_id=bundle_id,
```

Comment:
I think for new uploads with should default the `bundle_id` here (eg: generate a random UUID if none is supplied). Logic: that way we can then start encoding the bundle_id/debug_id into the URL for deleting artifact bundles that are not release bound.